qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java

```diff
@@ -332,84 +343,67 @@ public QuteTextDocument getDocument(String uri) {
 	 *         given function.
 	 */
 	public <R> CompletableFuture<R> getTemplate(TextDocumentIdentifier documentIdentifier,
-			BiFunction<CancelChecker, Template, R> code) {
-		return getTemplate(getDocument(documentIdentifier.getUri()), code);
+			BiFunction<Template, CancelChecker, R> code) {
+		return documents.computeModelAsync(documentIdentifier, code);
 	}
 
 	/**
-	 * Returns the properties model for a given uri in a future and then apply the
-	 * given function.
+	 * Parses the given Qute template file then passes the model to
+	 * the given function, then returns the result of the given function.
 	 *
-	 * @param <R>
-	 * @param documentIdentifier the document identifier.
-	 * @param code               a bi function that accepts a {@link CancelChecker}
-	 *                           and parsed {@link Template} and returns the to be
-	 *                           computed value
-	 * @return the properties model for a given uri in a future and then apply the
-	 *         given function.
+	 * Version of {@link TemplateFileTextDocumentService#getTemplate} that returns a
+	 * future of the value instead of the value itself
+	 *
+	 * @param <R>                The type of the result computed by the bifunction
+	 * @param documentIdentifier the document identifier
+	 * @param code               a bifunction that accepts the parsed
+	 *                           {@link Template} and
+	 *                           a {@link CancelChecker} and returns the value to be
+	 *                           computed as a future
+	 * @see {@link TemplateFileTextDocumentService#getTemplate}
+	 * @return the result of the passed bifunction as a future
 	 */
-	public <R> CompletableFuture<R> getTemplate(QuteTextDocument document,
-			BiFunction<CancelChecker, Template, R> code) {
-		return computeModelAsync(document.getModel(), code);
-	}
-
-	private static <R, M> CompletableFuture<R> computeModelAsync(CompletableFuture<M> loadModel,
-			BiFunction<CancelChecker, M, R> code) {
-		CompletableFuture<CancelChecker> start = new CompletableFuture<>();
-		CompletableFuture<R> result = start.thenCombineAsync(loadModel, code);
-		CancelChecker cancelIndicator = () -> {
-			if (result.isCancelled())
-				throw new CancellationException();
-		};
-		start.complete(cancelIndicator);
-		return result;
-	}
-
-	private static <R, M> CompletableFuture<R> computeModelAsync2(CompletableFuture<M> loadModel,
-			BiFunction<CancelChecker, M, CompletableFuture<R>> code) {
-		CompletableFuture<CancelChecker> start = new CompletableFuture<>();
-		CompletableFuture<R> result = start.thenCombineAsync(loadModel, code)//
-				.thenCompose(a -> {
-					return a;
-				});
-		CancelChecker cancelIndicator = () -> {
-			if (result.isCancelled())
-				throw new CancellationException();
-		};
-		start.complete(cancelIndicator);
-		return result;
+	public <R> CompletableFuture<R> getTemplateCompose(TextDocumentIdentifier documentIdentifier,

```

Comment:
It seems that LSP4MP has the same problem with findDefinition. Also I suggest that you create in the TextDocuments a new `computeModelAsync`  static method which does the compose with the following signature in `ModelTextDocuments`:

```java
public <R> CompletableFuture<R> computeModelAsyncCompose(TextDocumentIdentifier documentIdentifier,
			BiFunction<T, CancelChecker, CompletableFuture<R>> code) {
```

and if it works, we should change too LSPMP to use this new static method `ModelTextDocuments#computeModelAsyncCompose`

Reply:
I still need to make issues to copy it over to lsp4mp, lemminx, ... but I have made the change here